src/java/org/apache/cassandra/repair/autorepair/AutoRepairUtils.java

```diff
@@ -447,6 +453,182 @@ public static AutoRepairHistory getHostWithLongestUnrepairTime(RepairType repair
+            if (broadcastAddress == null)
+            {
+                logger.warn("Could not resolve broadcast address from host id {} in ClusterMetadata can't accurately " +
+                            "determine if this node is a replica of the local node.", hostId);
+            }
+            else
+            {
+                broadcastAddressMap.put(broadcastAddress, hostId);
+            }
+        }
+        return broadcastAddressMap;
+    }
+
+    /**
+     * @return Mapping of unique replication strategy to keyspaces using that strategy that we care about repairing.
+     */
+    private static Map<AbstractReplicationStrategy, List<String>> getReplicationStrategies()
+    {
+        // Collect all unique replication strategies among all keyspaces.
+        Map<AbstractReplicationStrategy, List<String>> replicationStrategies = new HashMap<>();
+        for (Keyspace keyspace : Keyspace.all())
+        {
+            if (AutoRepairUtils.shouldConsiderKeyspace(keyspace))
+            {
+                replicationStrategies.compute(keyspace.getReplicationStrategy(), (strategy, keyspaces) -> {
+                    if (keyspaces == null)
+                    {
+                        keyspaces = new ArrayList<>();
+                    }
+                    keyspaces.add(keyspace.getName());
+                    return keyspaces;
+                });
+            }
+        }
+        return replicationStrategies;
+    }
+
+    /**
+     * Identifies the most eligible host to repair for nodes preceding or equal to this nodes' lastRepairFinishTime.
+     * The criteria for this is to find the node with the oldest last repair finish time of which none of its replicasare currently under repair.
+     * @return The most eligible host to repair or null if no candidates before and including this nodes' current repair status.
+     */
+    @VisibleForTesting
+    public static AutoRepairHistory getMostEligibleHostToRepair(CurrentRepairStatus currentRepairStatus, UUID myId)
+    {
+        // 1. Sort repair histories from oldest completed to newest
+        Stream<AutoRepairHistory> finishedRepairHistories = currentRepairStatus.historiesWithoutOnGoingRepair
+                                                            .stream()
+                                                            .sorted((history0, history1) -> {
+                                                                if (history0.lastRepairFinishTime == history1.lastRepairFinishTime)
+                                                                    return 0;
+                                                                else
+                                                                    return history0.lastRepairFinishTime < history1.lastRepairFinishTime ? -1 : 1;
+                                                            });
+
+        // 2. Optimization: Truncate repair histories after myId so we don't evaluate anything more recent as if we
```

Comment:
I figured since we are only interested in determining if its this nodes turn and since its somewhat non-trivial work to determine if a node's replicas are being repaired that we short circuit the logic here.